Adding control for Tor Sharer options in Advanced Settings UI

nirvantyagi opened the pull request:
This pull request adds a component to the Advanced Settings UI that directly controls the Sharer's Tor settings. Namely, whether or not the Sharer would like to reproxy Getter's traffic through Tor.

It also adds a question to FAQ explaining how to use this feature. I've added a rough draft of what the text will look like, but will send out a draft for review and comments.

![advancedsettingstoroff](https://cloud.githubusercontent.com/assets/2855197/17073077/3db0740c-5022-11e6-8960-faf9bdd3b04a.png)

Advanced Settings page with Tor option off.

![advancedsettingtoron](https://cloud.githubusercontent.com/assets/2855197/17073081/491ec8e8-5022-11e6-8bc1-7c789587a175.png)

Advanced Settings page with Tor option on. The hidden port field is revealed.

![faqtor](https://cloud.githubusercontent.com/assets/2855197/17073092/5a683404-5022-11e6-8638-c6aaa8f9b2f5.png)

FAQ page.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2602)
<!-- Reviewable:end -->

bemasc commented on src/generic_ui/locales/en/messages.json:
```diff
@@ -171,6 +171,22 @@
+  },
+  "TOR_HOW_ENABLE_REPROXY_ANSWER": {
+    "description": "Title of FAQ question explaining Tor anonymity option of Sharers reproxying Getter requests",
+    "message": "<p><a href=\"https://www.torproject.org/\">Tor</a> is a free and open network that provides anonymity for its users by protecting against certain traffic analysis attacks. Though uProxy is not an anonymity tool by itself, it can be used with Tor to provide increased anonymity. Currently, our supported Tor option can only be enabled by the user sharing their access, not by users re ...
```
Nit: "anonmity".

I think the wording around "old users" is unclear.  Can we just say "all new connections"?  Or is this really a persistent per-user setting?

nirvantyagi replied:
An RtcToNet instance is created for each user upon the start of sharing taking in the global settings at that time. This instance then creates all future RtcToNet "sessions" for that user, meaning that changing the reproxy settings would not affect any "sessions" of this user. Thus, for new reproxy settings to take effect, the whole sharing instance for a user must be reset by revoking and reapproving.

Kind of a weird interaction, I agree. I could look into making the settings change take effect for future sessions, if you think that would be better.

bemasc commented:
This code looks fine, but I'm a bit worried that users might end up in a broken state if they turn on this feature but Tor is not running.  Is there anything we can do to avoid that situation?  If a user is in that broken state, do they see any error message that could help them get out of it?

nirvantyagi commented:
If a Sharer tries to use this feature without having Tor enabled, an error is caught when SOCKS auth is attempted with the Tor proxy port provided. Currently the error is just forwarded along as a Socks failure to the Getter, which results in "Site cannot be reached - ERR_SOCKS_CONNECTION_FAILED" for the Getter.

No notice to the Sharer, unless they are looking at the debug logs. I think it would be good to add a visual error to the Sharer if the SOCKS auth to the Tor proxy fails. I know where this error is caught; is there an example I can follow for how to display an error to the Sharer?

dborkan commented:
I wonder if maybe we should change "Enable TOR", to something like "Share through TOR"?  Enable TOR might mislead some getters to think that they are routing their traffic through TOR if they enable this (really it will have no effect for the getters).

Otherwise I think the UI code is good.  I agree we should look into better error messages in the failure case, but that could probably come in a separate pull request.

fortuna commented:
My suggestion would to:

1. Hide all Tor settings behind a single config entry
1. Display the setting status on a second line

Enabled example:
  **Tor Routing**
Routing through port 9050

Disabled example:
  **Tor Routing**
Not routing through Tor

Clicking on that takes you to a new page, with the Tor Settings:
**Enable Tor Routing**  [toggle]
**Tor Port**
9050

This is the pattern stock Android uses for settings.

nirvantyagi commented:
I will first merge the USERPASS auth pull request #2637. Then update with the relevant changes of adding auth options, before moving ahead with this pull request.

NOT READY TO MERGE

bemasc commented:
Review status: 0 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 389 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKZlfvBp6a3ioMQ9iy:-KPKZlfwO05JQPhiN52B:bbilyth) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_core/remote-connection.ts#L389)):*
> ```XML
>           break;
>         default:
>           throw new Error('Cannot handle status update');
> ```

IMHO this is scary; better to log a warning on an unrecognized status code.

---

*[src/generic_ui/locales/en/messages.json, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KNJkl_FkgizUoAKZneE-r1-188:-KPK_MW17zl6DfvNh9s0:b-njpkbn) ([raw file](https://github.com/uproxy/uproxy/blob/de3e8b756ae462cf98be16b37a3cf0e6774f30fb/src/generic_ui/locales/en/messages.json#L188)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> An RtcToNet instance is created for each user upon the start of sharing taking in the global settings at that time. This instance then creates all future RtcToNet "sessions" for that user, meaning that changing the reproxy settings would not affect any "sessions" of this user. Thus, for new reproxy settings to take effect, the whole sharing instance for a user must be reset by revoking and reapproving.
> 
> 
> 
> Kind of a weird interaction, I agree. I could look into making the settings change take effect for future sessions, if you think that would be better.
</details>
Yeah, but a "session" is just a socket.  Each new connection, at the time the getter clicks "Get Access", uses the latest reproxy settings.  So it's at connection-time, not permissioning-time.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 427 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKaIDu1APPVGvw3sDw:-KPKaIDu1APPVGvw3sDx:boh9iay) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/lib/rtc-to-net/rtc-to-net.ts#L427)):*
> ```XML
>               this.replyToPeer_(this.getReplyFromError_(e));
>               if (this.proxyConfig_.reproxy && this.proxyConfig_.reproxy.enabled) {
>                 this.statusUpdates_.handle(Status.REPROXY_ERROR);
> ```

Will this trigger if the getter tries to go to an invalid domain name?  That's not what we want; we only want an error if (a) we can't open a socket to the specified port, or (b) that port isn't running a SOCKS5 server.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Added two UI checks for the Sharer to help them understand if their reproxy is working.

1) On the Advanced Settings page, there is a check for whether a socks server is listening on the given port. It is automatically refreshed upon setting the reproxy settings, however it does not block the settings from being set.

2) Outside of the Advanced Settings page, if reproxy settings are enabled and rtc-to-net starts to fail to connect, a black error bar will appear and stay on the bottom of the screen until the reproxy is working again or reproxy settings are switched off.

Daniel noted - currently rtc-to-net sends updates for every working reproxy session connection. One optimization would be to store the state reproxyWorking in Session, and only send an update on a change.

1a Found server listening:
![server](https://cloud.githubusercontent.com/assets/2855197/17718859/74766d2c-63e4-11e6-9365-9df34e95503b.png)

1b Could not find server listening:
![no-socks](https://cloud.githubusercontent.com/assets/2855197/17718872/87d7ef08-63e4-11e6-911e-8d4e6d96662f.png)

2 Reproxy requests failing at rtc-to-net
![screenshot 2016-08-16 at 7 00 36 pm](https://cloud.githubusercontent.com/assets/2855197/17718880/94668d92-63e4-11e6-8720-13e70d989477.png)

nirvantyagi commented:
Review status: 0 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 427 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKaIDu1APPVGvw3sDw:-KPKbuGVU97o9o3O7Qaf:b-rbjhts) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/lib/rtc-to-net/rtc-to-net.ts#L427)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Will this trigger if the getter tries to go to an invalid domain name?  That's not what we want; we only want an error if (a) we can't open a socket to the specified port, or (b) that port isn't running a SOCKS5 server.
</details>
No, this only triggers if we cannot setup a TCP connection to the socksEndpoint given by the Sharer. So this update covers (a) and the update sent on line 440 covers (b)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602)*
<!-- Sent from Reviewable.io -->

bemasc commented on src/generic_core/uproxy_core.ts:
```diff
@@ -513,6 +515,26 @@ export class uProxyCore implements uproxy_core_api.CoreApi {
+        return uproxy_core_api.ReproxyCheck.TRUE;
+      }).catch((e :Error) :uproxy_core_api.ReproxyCheck => {
+        return uproxy_core_api.ReproxyCheck.FALSE;
+      });
```
Nothing closes `socksConnection`?   Leaks the socket.

bemasc commented:
Review status: 0 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/generic_core/remote-instance.ts, line 191 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKl-DjxUIGvEAyGS5L:-KPKl-DkJkX051mKFIq5:b4q3e4z) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_core/remote-instance.ts#L191)):*
> ```XML
>           this.user.notifyUI();
>           break;
>         case uproxy_core_api.Update.REPROXY_ERROR:
> ```

Nit:  This is a good place to use multiple case statements:
```typescript
case uproxy_core_api.Update.REPROXY_WORKING:
case uproxy_core_api.Update.REPROXY_ERROR:
  ui.update(update, data);
  break;
```

---

*[src/generic_ui/locales/en/messages.json, line 1433 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKlLcvMozIQd7abNMf:-KPKlLcvMozIQd7abNMg:b-p3rr4u) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/locales/en/messages.json#L1433)):*
> ```JSON
>   "REPROXY_ERR": {
>     "description": "Error message displayed when unable to route through reproxy",
>     "message": "Unable to route through reproxy"
> ```

How about something like: `"Sharing is broken because uProxy couldn't connect to Tor on port __PORT__.  Please start Tor or disable \"Share through Tor\" in the Advanced Settings."`

---

*[src/generic_ui/polymer/advanced-settings.ts, line 113 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKmBW9u6bt9Iyt-cIa:-KPKmBWAL5poNUTFxLwK:b-srrmm5) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/polymer/advanced-settings.ts#L113)):*
> ```XML
>     'opened': '$.advancedSettingsPanel.opened'
>   },
>   _reproxyCheckTrue: function(check: uproxy_core_api.ReproxyCheck) {
> ```

There must be a better way.  If you can't access the enum itself in the UI, maybe you could write a function that converts these values into four different strings?

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 590 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKmX4lUqmD1qMFAiqN:-KPKmX4lUqmD1qMFAiqO:b-96obcu) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/lib/rtc-to-net/rtc-to-net.ts#L590)):*
> ```XML
>         }).then(() :Promise<ArrayBuffer> => {
>           // Socks auth succeeded
>           this.statusUpdates_.handle(Status.REPROXY_WORKING);
> ```

This is the only place we report WORKING, right?  So if it's broken, and then the getter disconnects, are stuck with this notification in the UI forever?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 18 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 389 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKZlfvBp6a3ioMQ9iy:-KPPCLHBojw5_Pgu0RvE:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_core/remote-connection.ts#L389)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> IMHO this is scary; better to log a warning on an unrecognized status code.
</details>
Done.

---

*[src/generic_core/remote-instance.ts, line 191 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKl-DjxUIGvEAyGS5L:-KPPD2DydMc9brhrVGZK:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_core/remote-instance.ts#L191)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit:  This is a good place to use multiple case statements:
> ```typescript
> case uproxy_core_api.Update.REPROXY_WORKING:
> case uproxy_core_api.Update.REPROXY_ERROR:
>   ui.update(update, data);
>   break;
> ```
</details>
Done.

---

*[src/generic_core/uproxy_core.ts, line 535 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KOqUuqSsl2gxwu36JgB-r4-535:-KPPJ2a8l13PXp4ztPp9:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_core/uproxy_core.ts#L535)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nothing closes `socksConnection`?   Leaks the socket.
</details>
Done.

---

*[src/generic_ui/locales/en/messages.json, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KNJkl_FkgizUoAKZneE-r1-188:-KPPJLkbYzr3s6BRBqqE:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/de3e8b756ae462cf98be16b37a3cf0e6774f30fb/src/generic_ui/locales/en/messages.json#L188)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Yeah, but a "session" is just a socket.  Each new connection, at the time the getter clicks "Get Access", uses the latest reproxy settings.  So it's at connection-time, not permissioning-time.
</details>
Done.

---

*[src/generic_ui/locales/en/messages.json, line 1433 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKlLcvMozIQd7abNMf:-KPQ97uJ7bMNxwuXFg_U:b-xtoh0n) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/locales/en/messages.json#L1433)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> How about something like: `"Sharing is broken because uProxy couldn't connect to Tor on port __PORT__.  Please start Tor or disable \"Share through Tor\" in the Advanced Settings."`
</details>
Done. Added a dialog box pop-up on 'Get Help' link same as the black bar warning does for Proxy Error.

---

*[src/generic_ui/polymer/advanced-settings.ts, line 113 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKmBW9u6bt9Iyt-cIa:-KPPURGd6DASHOqLXhnC:b-fazlh) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/polymer/advanced-settings.ts#L113)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> There must be a better way.  If you can't access the enum itself in the UI, maybe you could write a function that converts these values into four different strings?
</details>
Can't access enum, but added fxn to convert to strings.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 590 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKmX4lUqmD1qMFAiqN:-KPPWM0N-pjgtPh7CEm2:bilyxum) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/lib/rtc-to-net/rtc-to-net.ts#L590)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This is the only place we report WORKING, right?  So if it's broken, and then the getter disconnects, are stuck with this notification in the UI forever?
</details>
It is reset to "working" when the Sharer turns off reproxy in Advanced Settings. But yeah...if the Sharer keeps reproxying on, but has no Getters, then currently it will never remove the notification even if it is fixed at some point.

UPDATE: Fixed by adding background interval to checkReproxy and update value even when there is no active Getter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602)*
<!-- Sent from Reviewable.io -->

bemasc commented:
:+1: with wording nits

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/generic_ui/locales/en/messages.json, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KNJkl_FkgizUoAKZneE-r1-188:-KPT5WvVnqbvr1-XYPNj:b-vuva92) ([raw file](https://github.com/uproxy/uproxy/blob/de3e8b756ae462cf98be16b37a3cf0e6774f30fb/src/generic_ui/locales/en/messages.json#L188)):*
> All new connections you grant access to will now have their requests routed through Tor.

Nit: remove "you grant access to".

> To send traffic from previously approved connections through Tor, you must revoke and reapprove their access.

I think you can just drop this sentence.

---

*[src/generic_ui/locales/en/messages.json, line 1433 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKlLcvMozIQd7abNMf:-KPT6H6LaxuhACF9_6RC:bfblkj4) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/locales/en/messages.json#L1433)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Done. Added a dialog box pop-up on 'Get Help' link same as the black bar warning does for Proxy Error.
</details>
Why are REPROXY_ERR and REPROXY_ERROR_HEADING the same?  Could we just have one?

---

*[src/generic_ui/locales/en/messages.json, line 1441 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPT6LezhunHf7BcyyTQ:-KPT6LezhunHf7BcyyTR:b-6d0bhk) ([raw file](https://github.com/uproxy/uproxy/blob/9ecf19498c6054fac565429173ab5ac362e6d7c7/src/generic_ui/locales/en/messages.json#L1441)):*
> ```JSON
>   "REPROXY_ERROR_DETAILS": {
>     "description": "Body of expanded reproxy error dialog box",
>     "message": "Sharing is broken because uProxy couldn't connect to Tor on input port. Please start Tor or disable \"Share through Tor\" in Advanced Settings."
> ```

Nit: "Tor on input port" -> "Tor's input port".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-:-KPT7V3Y1ZqPSxWTbEqr:b-f75kbp)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 18 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[src/generic_ui/locales/en/messages.json, line 188 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KNJkl_FkgizUoAKZneE-r1-188:-KPTMxC7AkKyWqF-TPhw:b2v1kve) ([raw file](https://github.com/uproxy/uproxy/blob/de3e8b756ae462cf98be16b37a3cf0e6774f30fb/src/generic_ui/locales/en/messages.json#L188)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> > All new connections you grant access to will now have their requests routed through Tor.
> 
> Nit: remove "you grant access to".
> 
> > To send traffic from previously approved connections through Tor, you must revoke and reapprove their access.
> 
> I think you can just drop this sentence.
</details>
Done. I will make sure this interaction is clear in the blog post.

---

*[src/generic_ui/locales/en/messages.json, line 1433 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPKlLcvMozIQd7abNMf:-KPTMXGZ9duPdw3lmJ65:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/87755efc3ee89cc3e9bd206174c87190bd7e6f8d/src/generic_ui/locales/en/messages.json#L1433)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Why are REPROXY_ERR and REPROXY_ERROR_HEADING the same?  Could we just have one?
</details>
Done.

---

*[src/generic_ui/locales/en/messages.json, line 1441 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-KPT6LezhunHf7BcyyTQ:-KPTMhfqwWRZFxuk8wVJ:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/9ecf19498c6054fac565429173ab5ac362e6d7c7/src/generic_ui/locales/en/messages.json#L1441)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: "Tor on input port" -> "Tor's input port".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602)*
<!-- Sent from Reviewable.io -->

bemasc commented:
:thumbsup:

---

Review status: 0 of 18 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2602#-:-KPTPSeN58KBwRLtzWJe:bn1wz0t)*
<!-- Sent from Reviewable.io -->
